### PR TITLE
[MLIR][Runtime] Change `__quantum__` to` __catalyst__` for all catalyst defined library.

### PR DESCRIPTION
**Pull request by erick-xanadu**
**Context:** QIR defines an intermediate representation for quantum operations. This is done by defining quantum operations as function calls in LLVM-IR. It is possible to implement quantum operations as function calls, but this is not the only way to generate a valid executable. Due to QIR defining quantum operations as functions, it is very natural to simply provide a definition for these functions and IR with the library defining these functions. However, if two different implementers implemented the same functions, then the process would have issues finding the correct location. There are a couple of ways to solve this, but one quick solution would just be to use namespaces.

Namespaces in C++ are implemented by namemangling. Since QIR's functions are C like, there is no name mangling. But we can create a namespace by just preppending by using a unique name.

**Description of the Change:** Replaces `__quantum__` with `__catalyst__` in the runtime and generated code for QIR "functions" we override.

**Benefits:** A program can be linked against catalyst and a second implementer.

**Possible Drawbacks:** This PR does not rename of all `__quantum__` functions. We depend on some QIR defined functions. We might want to avoid relying on other QIR runner libraries to avoid symbol conflicts if third party libraries redefine these functions. This is not necessarily a drawback, since it will still be present regardless of whether this PR is merged or not. But I am pointing that a more complete implementation of this PR would also get rid of QIR runner library as a dependency.

**Related GitHub Issues:**

[sc-55563]

### CODE REVIEW

**Comment by maliasadi** [comment]
@erick-xanadu Is this PR unblocking any story? How's it related to our roadmap and what is this second implementer?

**Comment by dime10** [comment]
After some thought I actually think this is a pretty decent solution, especially if we can couple this to removing the qir runtime library from our build system altogether.

**Review by maliasadi — APPROVED** [review]
Thanks @erick-xanadu! looks really good from my side :tada: 

Don't forget to update changelog as this affects our public C-API.

**Comment by maliasadi** [comment]
Have you ever got time to check the symbol versioning as an alternative approach? 

``` c
void __quantum__rt__initialize() __attribute__((version("CATALYST_1.0")));
```

**Comment by erick-xanadu** [comment]
> void __quantum__rt__initialize() __attribute__((version("CATALYST_1.0")));

I am having a little bit of trouble coming up with how to use this in our own library. From this example: https://man7.org/conf/lca2006/shared_libraries/slide19a.html we need two different implementations of `__quantum__rt__initialize` in our library. We only have one definition.

I do have another alternative which is we compile all of the runtime into LLVM bytecode, ship the bytecode (as opposed to the compiled object), and then link them all together into a single module and mark all functions definitions as private. This would increase compilation time, but I like this since it removes any influence from the environment. This is equivalent to statically linking the runtime against the generated object.

File: doc/changelog.md [hunk 1]
```diff
@@ -15,6 +15,33 @@
...
+  the `__quantum__` part of the function name is replaced with `__catalyst__`. This prevents
+  the possibility of symbol conflicts with other libraries that implement QIR as a library.
+  However, it doesn't solve it completely. Since the `__catalyst__` functions are still exported
+  if another library implemented the same symbols exported by the runtime, the same problem would
+  presist.
+
+  Another potential way to solve this is for the compiler to generate code for the user's function
+  without referring to any external symbol. This way, the generated code would only refer to symbols
+  inside the translation unit itself. This alternative approach would increase the compilation time.
+  The approach in this commit may be revisited in the future.
```
Comment (dime10) [inline doc/changelog.md:43]: Maybe better to move this aspect to the PR description,  as it's not relevant to the changelog?

**Review by dime10 — APPROVED** [review]
Good to go I would say 💯

File: doc/changelog.md (hunk 1 above)
Comment (erick-xanadu) [reply doc/changelog.md]: Removed it here: https://github.com/PennyLaneAI/catalyst/pull/464/commits/92629968fac8c3020e0636f75a4279e8154a2a52 Thanks!

**Pull request merged**